### PR TITLE
Consider a runner with disabled redis services

### LINKED ISSUE #524: Enable plugin writers to disable the Redis service for specific test runners

### PR DESCRIPTION
**Pull request by lubosmj**
closes #524

### CODE REVIEW

**Comment by lubosmj** [comment]
This commit might be useful across all the plugins since we probably want to test the pulp services without the Redis server. Modifying the `script.sh` is not recommended by default (https://github.com/pulp/pulpcore/pull/1663/files#diff-02059c243ab1ae5faa6a95ec965f0b351e0f6c96e5452c90881b6b7e3ee11618). At the moment, this change enables setting up the configuration for the fixtures.

I am thinking about introducing a new option to the template config. Not sure if it will be that needed, though.

**Comment by mdellweg** [comment]
I would prefer, if we could fold this into on of s3 or azure tests. It should be reasonably independent to not need another full excercise of the whole test suite.

**Comment by dralley** [comment]
I agree with @mdellweg

**Comment by lubosmj** [comment]
I have introduced a new configuration option. Please take a look at the changes, @mdellweg.

File: .github/workflows/bootstrap.yaml [hunk 1]
```diff
@@ -72,6 +72,8 @@ jobs:
     runs-on: ubuntu-latest
     # run only after lint finishes
     needs: lint
+    env:
+      REDIS_DISABLED: "azure s3"
```
Comment (dralley) [inline .github/workflows/bootstrap.yaml:76]: Don't disable caching for both, we still want some coverage of a non-local storage backend in combination with the cache.

File: templates/github/.github/workflows/ci.yml.j2 [hunk 2]
```diff
@@ -91,6 +91,12 @@ jobs:
     runs-on: ubuntu-latest
     # run only after lint finishes
     needs: lint
+    env:
+    {%- if disabled_redis_runners %}
+      REDIS_DISABLED: "{{ disabled_redis_runners | join(' ') }}"
+    {%- else %}
+      REDIS_DISABLED: ""
```
Comment (mdellweg) [inline templates/github/.github/workflows/ci.yml.j2:98]: I'd rather have a `REDIS_DISABLED=true/false` a few lines down in the matrix.

Some thing like maybe:
`REDIS_DISABLED: "{{ "s3" in (disabled_redis_runners | default([])) }}"`

I think, this makes the generated workflows much more readable. Maybe it's also possible outside of the matrix with
```
env:
  - REDIS_DISABLED: ${{ matrix.env.TEST == "s3" }}
```
This way the workflows would keep their current names.

File: templates/github/.github/workflows/scripts/before_script.sh.j2 [hunk 3]
```diff
@@ -29,6 +29,14 @@ if [[ "$TEST" == 'pulp' || "$TEST" == 'performance' || "$TEST" == 'upgrade' || "
   cmd_prefix dnf install -yq lsof which dnf-plugins-core
 fi
 
+{%- if disabled_redis_runners %}
+read -ra REDIS_DISABLED <<< "$REDIS_DISABLED"
+if [[ " ${REDIS_DISABLED[*]} " =~ " ${TEST} " ]]; then
+  cmd_prefix bash -c "s6-svc -d /var/run/s6/services/redis"
+  echo "The Redis service was disabled for $TEST"
+fi
+{%- endif %}
```
Comment (mdellweg) [inline templates/github/.github/workflows/scripts/before_script.sh.j2:38]: Do we have a way to disable Redis also? Or is Pulp just able to fallback if missing?

Comment (lubosmj) [reply templates/github/.github/workflows/scripts/before_script.sh.j2:38]: This disables the Redis server service. What do you mean by `disable Redis also`? The idea, that comes from https://pulp.plan.io/issues/8997, was to simulate the Pulp's behaviour after disabling the Redis service. The actual change to the template allows plugin writers to test their plugins in such a state as well.

File: templates/github/.github/workflows/ci.yml.j2 (hunk 2 above)
Comment (lubosmj) [reply templates/github/.github/workflows/ci.yml.j2:98]: I do not know where else a few lines down in the matrix you could define an environment variable. I think I cannot do this:
```
    strategy:
      fail-fast: false
      matrix:
        env:
          - TEST: pulp
          - TEST: docs
          - TEST: azure
            - REDIS_DISABLED: false
          - TEST: s3
            - REDIS_DISABLED: true
```

Comment (mdellweg) [reply templates/github/.github/workflows/ci.yml.j2:98]: you could do:
```
    strategy:
      fail-fast: false
      matrix:
        env:
          - TEST: pulp
          - TEST: docs
          - TEST: azure
            REDIS_DISABLED: false
          - TEST: s3
            REDIS_DISABLED: true
```
But on a second thought, this changes the names of the workflow jobs and that'd need adjusting the repo settings everywhere.
So i'd really go with an `env` stanza in the `steps:` part of the workflow.

File: templates/github/.github/workflows/scripts/before_script.sh.j2 (hunk 3 above)
Comment (mdellweg) [reply templates/github/.github/workflows/scripts/before_script.sh.j2:38]: I'd say this shuts the service down, simulating a failure. But what if a user wants to run Pulp without ever installing / configuring redis. Would this use different codepaths in any way? Do we not support such a thing yet?

Comment (lubosmj) [reply templates/github/.github/workflows/scripts/before_script.sh.j2:38]: There is a separate issue filed for that: https://pulp.plan.io/issues/9475. Should I rather focus on that one in this PR instead?

Comment (mdellweg) [reply templates/github/.github/workflows/scripts/before_script.sh.j2:38]: No, just wanted to know if we miss an opportunity today.

**Comment by lubosmj** [comment]
@mdellweg, I made the changes you have proposed. Now, the environment variable `REDIS_DISABLED` is defined in the place where all other environment variables reside for each runner.

File: plugin-template [hunk 4]
```diff
@@ -66,6 +66,7 @@ DEFAULT_SETTINGS = {
     "test_cli": False,
     "test_performance": False,
     "test_released_plugin_with_next_pulpcore_release": False,
+    "disabled_redis_runners": [],
```
Comment (mdellweg) [inline plugin-template:69]: Wouldn't it be the easiest to just default to "s3" here? It shouldn't fail if "s3" wasn't even tested, no?

File: templates/github/.github/workflows/scripts/before_script.sh.j2 [hunk 5]
```diff
@@ -29,6 +29,11 @@ if [[ "$TEST" == 'pulp' || "$TEST" == 'performance' || "$TEST" == 'upgrade' || "
   cmd_prefix dnf install -yq lsof which dnf-plugins-core
 fi
 
+if [[ "$REDIS_DISABLED" == true ]]; then
+  cmd_prefix bash -c "s6-svc -d /var/run/s6/services/redis"
+  echo "The Redis service was disabled for $TEST"
+fi
```
Comment (mdellweg) [inline templates/github/.github/workflows/scripts/before_script.sh.j2:35]: This is nice and readable!

File: .github/workflows/bootstrap.yaml [hunk 6]
```diff
@@ -123,8 +123,10 @@ jobs:
         run: .github/workflows/scripts/install_python_client.sh
 
       - name: Before Script
-        run: |
-          .github/workflows/scripts/before_script.sh
+        run: .github/workflows/scripts/before_script.sh
+        shell: bash
+        env:
+          REDIS_DISABLED: ${{ contains('s3', matrix.env.TEST) }}
```
Comment (mdellweg) [inline .github/workflows/bootstrap.yaml:129]: I sense that this is not exactly the same as in the file that gets generated. (Not your fault!) Can you regenerate it please?
@fao do you see a way that we can asure this part of the file equals the on the fly generated bit by bit?

File: templates/macros.j2 [hunk 7]
```diff
@@ -81,14 +86,14 @@ GITHUB_CONTEXT: {{ "${{ github.event.pull_request.commits_url }}" }}
 {%- endmacro -%}
 
 
-{%- macro run_script(name, file, withenv=True, condition=None) -%}
+{%- macro run_script(name, file, withenv=True, condition=None, customenvs=None) -%}
```
Comment (mdellweg) [inline templates/macros.j2:89]: I would name that variable `extra_env` or just `env`.

File: plugin-template (hunk 4 above)
Comment (lubosmj) [reply plugin-template:69]: It should not fail. Yet, I think that we should not disable Redis for any storage by default.

File: .github/workflows/bootstrap.yaml (hunk 6 above)
Comment (fao89) [reply .github/workflows/bootstrap.yaml:129]: I saw this tweet yesterday:
https://twitter.com/github/status/1461106904122417156?s=20
about reusable workflows, I didn't read the article, but maybe this is the solution

Comment (lubosmj) [reply .github/workflows/bootstrap.yaml:129]: No, the question was about comparing two files on the fly. We are not going to run another workflow right now.

File: templates/github/.github/workflows/scripts/before_script.sh.j2 [hunk 8]
```diff
@@ -29,6 +29,11 @@ if [[ "$TEST" == 'pulp' || "$TEST" == 'performance' || "$TEST" == 'upgrade' || "
   cmd_prefix dnf install -yq lsof which dnf-plugins-core
 fi
 
+if [[ "$REDIS_DISABLED" == true ]]; then
```
Comment (mdellweg) [inline templates/github/.github/workflows/scripts/before_script.sh.j2:32]: This fails here:
https://github.com/pulp/pulpcore/runs/4319788337?check_suite_focus=true#step:12:143

```suggestion
if [[ "${REDIS_DISABLED:-false}" == true ]]; then
```

**Review by mdellweg — APPROVED** [review]

**Pull request merged**